README.md
# nowa-server

[![NPM version](https://img.shields.io/npm/v/nowa-server.svg?style=flat)](https://npmjs.org/package/nowa-server)

nowa webpack dev server

---

## Feature

- nowa webpack dev server

## Install

```bash
$ npm i nowa -g
$ nowa install server
```

## Usage

```bash
$ nowa server --help
```


CHANGELOG.md
# Changelog

## 1.19.0 (2017-11-17)

* [feature] Remove include restriction for several loader.

> https://github.com/nowa-webpack/nowa-server/pull/20

## 1.18.0 (2017-09-05)

* [feature] Update version of ts-loader.

## 1.17.0 (2017-06-21)

* [feature] Support IE<=10 `static extends` on `loose` mode.

## 1.16.0 (2017-05-18)

* [feature] Inject `webpack` to user defined config.

## 1.15.3 (2017-05-11)

* [fix] Revert to 1.15.0.

## 1.15.1 (2017-05-10)

* [fix] .ts files support jsx.

## 1.15.0 (2017-04-21)

* [feature] Add `--no-sourcemap` option to disable sourcemap.

> https://github.com/nowa-webpack/nowa-server/pull/19

## 1.14.1 (2017-03-20)

* [fix] Update react version to resolve peerDependencies issue.

## 1.14.0 (2017-03-16)

* [feature] Proxy support rewrite path.

> https://github.com/nowa-webpack/nowa/issues/39

## 1.13.0 (2017-03-16)

* [feature] Add transform-decorators plugin.

> https://github.com/nowa-webpack/nowa/issues/38

## 1.12.0 (2017-03-12)

* [feature] Support page filter to add specified pages to entry.

> https://github.com/nowa-webpack/nowa/issues/35

## 1.11.1 (2017-03-09)

* [feature] Add process info.

> https://github.com/nowa-webpack/nowa-server/issues/18

## 1.10.3 (2017-02-09)

* [improve] Add inject point.

> https://github.com/nowa-webpack/nowa-server/pull/16

## 1.10.2 (2017-01-23)

* [improve] Improve injection algorithm.

## 1.10.1 (2017-01-23)

* [feature] Support html injection for css resources.

## 1.10.0 (2017-01-19)

* [feature] Support html injection.
* [feature] Add host option to support custom host.
* [feature] Add TypeScript support.

> https://github.com/nowa-webpack/nowa-server/issues/14

> https://github.com/nowa-webpack/nowa-server/pull/7

> https://github.com/nowa-webpack/nowa-server/pull/15

## 1.9.1 (2017-01-17)

* [fix] Fix loader found error issue.

> https://github.com/nowa-webpack/nowa-server/issues/11

## 1.9.0 (2016-12-26)

* [fix] Fix root dependency mismatch bug.

> https://github.com/nowa-webpack/nowa/issues/30

## 1.8.3 (2016-12-22)

* [fix] Update less version.

## 1.8.2 (2016-12-22)

* [fix] Update stylus-loader version for supporting Nodejs 6+.

## 1.8.1 (2016-12-13)

* [fix] Fix alias bug.

## 1.8.0 (2016-12-11)

* [feature] Add alias option for module path alias defination.

> https://github.com/nowa-webpack/nowa-server/issues/10

## 1.7.0 (2016-11-22)

* [feature] Add jsx extension support.

> https://github.com/nowa-webpack/nowa-server/pull/9

## 1.6.1 (2016-11-17)

* [fix] Add missing babel-runtime for non-lazyload mode.

> https://github.com/nowa-webpack/nowa-server/issues/8

## 1.6.0 (2016-10-23)

* [feature] Support font files(woff,woff2,ttf,otf).

## 1.5.0 (2016-10-14)

* [feature] Use eval mode to build sourcemap.

## 1.4.1 (2016-09-27)

* [fix] babel-runtime do not support old IEs, so use es3ify-loader to transform babel-runtime.

> https://github.com/nowa-webpack/nowa/issues/22

## 1.4.0 (2016-09-27)

* [feature] Update version of babel plugins.
* [feature] Use es2015 internel loose mode.
* [feature] Use babel-runtime to support polyfill.

> https://github.com/nowa-webpack/nowa-build/issues/10

> https://github.com/nowa-webpack/nowa/issues/20

## 1.3.0 (2016-08-23)

* [feature] Add include option to support adding more include path for loader.

> https://github.com/nowa-webpack/nowa-build/issues/9

## 1.2.0 (2016-07-27)

* [feature] Ignore babelrc.
* [improve] Add cache directory for babel to speed up.

## 1.1.2 (2016-07-12)

* [fix] `--loose` argument do not take effect.

> https://github.com/nowa-webpack/nowa-server/issues/4

## 1.1.1 (2016-06-30)

* [fix] Can not find nowa plugin when nowa was locally installed.

> https://github.com/nowa-webpack/nowa/issues/6

## 1.1.0 (2016-06-27)

* [feature] Support varible replacement in historyApiFallback.

> https://github.com/nowa-webpack/nowa-server/issues/3

## 1.0.0 (2016-06-26)

* [feature] Add historyApiFallback.
* [feature] Add mockapi.

> https://github.com/nowa-webpack/nowa-server/issues/1

> https://github.com/nowa-webpack/nowa-server/issues/2


src/util.js
/*
* @Author: gbk
* @Date:   2016-05-02 17:15:36
* @Last Modified by:   gbk
* @Last Modified time: 2017-03-12 16:15:32
*/

'use strict';

var os = require('os');
var fs = require('fs');
var path = require('path');
var spawn = require('child_process').spawn;

var util = {

  // get absolute path to cwd
  cwdPath: function() {
    var argvs = Array.prototype.slice.call(arguments);
    argvs.unshift(process.cwd());
    return path.join.apply(path, argvs);
  },

  // get absolute path to __dirname
  relPath: function(p) {
    var argvs = Array.prototype.slice.call(arguments);
    argvs.unshift(__dirname);
    return path.join.apply(path, argvs);
  },

  // make a webpack entry
  makeEntry: function(options) {
    var params = Array.prototype.slice.call(arguments, 1);
    if (options.lazyload) {
      return './' + path.join.apply(path, params);
    } else {
      var address = encodeURIComponent(options.address + '/__webpack_hmr');
      return [
        'webpack-hot-middleware/client?reload=true&noInfo=true&path=' + address,
        './' + path.join.apply(path, params)
      ]
    }
  },

  // make all valid pages as webpack entries
  makePageEntries: function(options, src, entries, pagesFilter) {
    var pages = fs.readdirSync(path.join(src, 'pages'));
    if (typeof pagesFilter === 'string') {
      pagesFilter = pagesFilter.split(',');
      pages = pages.filter(function(page) {
        return pagesFilter.indexOf(page) !== -1;
      });
    }
    pages.forEach(function(page) {
      try {
        var entry = path.join(src, 'pages', page, 'index.js');
        if (fs.statSync(entry).isFile()) {
          entries[page] = util.makeEntry(options, src, 'pages', page, 'index.js');
          return;
        }
      } catch (e) {
      }
      try {
        var entry = path.join(src, 'pages', page, 'index.jsx');
        if (fs.statSync(entry).isFile()) {
          entries[page] = util.makeEntry(options, src, 'pages', page, 'index.jsx');
        }
      } catch (e) {
      }
    });
    return entries;
  },

  // parse vars for DefinePlugin
  parseVars: function(vars) {
    var newVars = {};
    for (var key in vars) {
      newVars[key] = JSON.stringify(vars[key]);
    }
    return newVars;
  },

  // make filename suffix by vars
  suffixByVars: function (vars, buildvars) {
    if (vars) {
      var suffix = '';
      for (var key in vars) {
        var value = vars[key];

        // filename suffix will not contain `/`
        if (value !== undefined && buildvars && buildvars[key] && buildvars[key].length > 1) {
          suffix += '-' + value.toString().replace(/\//, '');
        }
      }
      return suffix;
    } else {
      return '';
    }
  },

  // make babel plugin/preset absolute path
  babel: function(type, name) {
    if (Array.isArray(name)) {
      return name.map(function(n) {
        return util.babel(type, n);
      });
    } else {
      if (typeof name === 'object') {
        return [
          require.resolve([
            'babel',
            type,
            name.name
          ].join('-')),
          name.options
        ];
      } else {
        return require.resolve([
          'babel',
          type,
          name
        ].join('-'));
      }
    }
  },

  // safe watch
  watch: function(dir, handler) {
    var timer = 0;
    fs.watch(dir, function() {
      clearTimeout(timer);
      timer = setTimeout(handler, 20);
    });
  },

  // make options for https key & cert
  makeKeyCert: function(httpsOpt) {

    if (httpsOpt === true) {

      // use preset key & cert
      return {
        key: fs.readFileSync(util.relPath('certs', 'host.key')),
        cert: fs.readFileSync(util.relPath('certs', 'host.cert'))
      };
    } else {

      // user passed key & cert
      return {
        key: fs.readFileSync(httpsOpt.key),
        cert: fs.readFileSync(httpsOpt.cert)
      };
    }
  },

  // open url in default browser
  open: function(url) {
    var command;
    switch (process.platform) {
      case 'darwin':
        command = 'open';
        break;
      case 'win32':
        command = 'explorer.exe';
        break;
      case 'linux':
        command = 'xdg-open';
        break;
      default:
        console.error('Can not open browser');
        return;
    }
    spawn(command, [ url ]);
  },

  // test if the path is a directory
  isDirectory: function(file) {
    try {
      return fs.statSync(file).isDirectory();
    } catch (e) {
      return false;
    }
  },

  // write process info
  writeProcessInfo: function(info) {
    try {
      var infoPath = path.join(os.tmpdir(), '.nowa-server-' + info.uid + '.json');
      fs.writeFileSync(infoPath, JSON.stringify(info, null, '  '));
      return info;
    } catch (e) {
      return null;
    }

  },
};

module.exports = util;


src/loader.js
/*
* @Author: gbk
* @Date:   2016-05-02 22:07:46
* @Last Modified by:   gbk
* @Last Modified time: 2017-06-21 20:00:57
*/

'use strict';

var os = require('os');
var path = require('path');

var util = require('./util');

module.exports = function(options) {
  var srcPath = util.cwdPath(options.src);
  if (options.includes) {
    srcPath = [
      srcPath
    ].concat(options.includes.map(function(include) {
      return util.cwdPath(include);
    }));
  }
  var preLoader = options.keepcss ? 'style-loader!export-css-loader?remove=false' : 'style-loader';
  var presets = util.babel('preset', [
    {
      name: 'es2015',
      options: {
        loose: !!options.loose
      }
    },
    'stage-0',
    'react'
  ]);
  var cacheDirectory = path.join(os.tmpdir(), options.loose ? 'babel-loose' : 'babel-strict');
  var plugins = [
    'add-module-exports',
    'transform-decorators-legacy',
    'transform-es3-member-expression-literals',
    'transform-es3-property-literals',
    {
      name: 'transform-runtime',
      options: {
        polyfill: !!options.polyfill,
        helpers: false,
        regenerator: true
      }
    }
  ];
  if (options.loose) {
    plugins.push('transform-proto-to-assign');
  }
  return [{
    test: /\.jsx?$/,
    loader: 'babel-loader',
    include: srcPath,
    query: options.lazyload ? {
      plugins: util.babel('plugin', plugins),
      presets: presets,
      cacheDirectory: cacheDirectory,
      babelrc: false
    } : {
      plugins: util.babel('plugin', [
        'add-module-exports',
        'transform-decorators-legacy',
        {
          name: 'transform-runtime',
          options: {
            polyfill: !!options.polyfill,
            helpers: false,
            regenerator: true
          }
        },
        {
          name: 'react-transform',
          options: {
            transforms: [
              {
                transform: 'react-transform-hmr',
                imports: [ 'react' ],
                locals: [ 'module' ]
              }, {
                transform: 'react-transform-catch-errors',
                imports: [ 'react', 'redbox-react' ]
              }
            ]
          }
        }
      ]),
      presets: presets,
      cacheDirectory: cacheDirectory,
      babelrc: false
    }
  }, {
    test: /\.js$/,
    loader: 'es3ify-loader',
    include: function(path) {
      return ~path.indexOf('babel-runtime');
    }
  }, {
    test: /\.css$/,
    loader: preLoader + '!css-loader?sourceMap',
  }, {
    test: /\.less$/,
    loader: preLoader + '!css-loader?sourceMap!less-loader',
  }, {
    test: /\.styl$/,
    loader: preLoader + '!css-loader?sourceMap!stylus-loader',
  }, {
    test: /\.svg$/,
    loader: 'babel-loader',
    query: {
      presets: presets,
      cacheDirectory: cacheDirectory,
      babelrc: false
    }
  }, {
    test: /\.svg$/,
    loader: 'svg2react-loader',
  }, {
    test: /\.json$/,
    loader: 'json-loader',
  }, {
    test: /\.(png|jpe?g|gif|woff|woff2|ttf|otf)$/,
    loader: 'url-loader',
  },{
    test: /\.tsx?$/,
    loader: 'ts-loader',
    include: srcPath,
    query: {
      transpileOnly: true
    }
  }];
};


src/static.js
/*
* @Author: gbk
* @Date:   2016-03-28 16:42:29
* @Last Modified by:   gbk
* @Last Modified time: 2017-01-19 20:39:39
*/

'use strict';

var fs = require('fs');
var path = require('path');

var parseUrl = require('parseurl');
var send = require('send');

var util = require('./util');

// simple version for express.static
// thanks to https://github.com/expressjs/serve-static
module.exports = function(root, opts) {
  opts = opts || {};
  opts.root = path.resolve(root);

  if (util.isDirectory(opts.root)) {

    // root must be a directory
    return function(req, res, next) {
      var forwardError = false;
      send(req, parseUrl(req).pathname, opts)
        .on('directory', function() {
          this.error(404);
        })
        .on('file', function() {
          forwardError = true;
        })
        .on('error', function(err) {
          if (forwardError || !(err.statusCode < 500)) {
            next(err);
            return;
          }
          next();
        })
        .pipe(res);
    };

  } else {

    // is not, do nothing
    return function(req, res, next) {
      next();
    };

  }
};


src/injects.js
/*
* @Author: gbk
* @Date:   2017-01-19 20:20:58
* @Last Modified by:   Tommy Troy Lin
* @Last Modified time: 2017-02-08 18:06:34
*/

'use strict';

var fs = require('fs');
var path = require('path');

var parseUrl = require('parseurl');
var resolveCwd = require('resolve-cwd');

var util = require('./util');

module.exports = function (root, injects) {
  root = path.resolve(root);
  injects = (injects || []).map(function (inject) {
    var res = (resolveCwd(inject) || inject).replace(process.cwd(), '');
    if (/\.css$/.test(res)) {
      return '<link rel="stylesheet" href="' + res + '" />';
    }
    return '<script src="' + res + '"></script>';
  }).join('');

  if (util.isDirectory(root) && injects.length) {

    // root must be a directory
    return function (req, res, next) {
      var pathname = parseUrl(req).pathname;
      if (pathname.substr(-1) === '/') {
        pathname += 'index.html';
      }
      if (/\.html?$/.test(pathname)) {
        var file = path.join(root, pathname);
        try {
          var content = fs.readFileSync(file).toString();
          var injectionComment = /<!--\s*Nowa\s*Server\s*Injects\s*-->/i;
          if (injectionComment.test(content)) {
            content = content.replace(injectionComment, injects);
          }
          res.end(content);
        } catch (e) {
          next();
        }
      } else {
        next();
      }
    };

  } else {

    // is not, do nothing
    return function (req, res, next) {
      next();
    };

  }
};


src/index.js
/*
* @Author: gbk <ck0123456@gmail.com>
* @Date:   2016-04-21 17:34:00
* @Last Modified by:   gbk
* @Last Modified time: 2017-05-18 00:32:03
*/

'use strict';

var url = require('url');
var https = require('https');

var express = require('express');
var webpack = require('webpack');
var webpackDevMiddleware = require('webpack-dev-middleware');
var webpackHotMiddleware = require('webpack-hot-middleware');
var chalk = require('chalk');
var ip = require('ip');
var portscanner = require('portscanner');
var httpProxy = require('express-http-proxy');

var util = require('./util');
var loader = require('./loader');
var favicon = require('./favicon');
var serveStatic = require('./static');
var injectsHtml = require('./injects');
var pkg = require('../package.json');

// plugin defination
module.exports = {

  description: pkg.description,

  options: [
    [ '-s, --src <dir>', 'source directory, default to `src`', 'src' ],
    [ '-d, --dist <dir>', 'build directory, default to `dist`', 'dist' ],
    [ '    --host <host>', 'server host default to first private ip'],
    [ '-p, --port <port>', 'server port, default to `3000`', 3000 ],
    [ '-e  --entry <file>', 'app entry, default to `app/app.js`', 'app/app.js' ],
    [ '    --pages [pages]', 'add multi-page entries' ],
    [ '    --vars', 'runtime context varibles' ],
    [ '    --buildvars', 'build varibles' ],
    [ '-r, --proxy <proxy>', 'dev proxy hostname or mappings' ],
    [ '-k, --keepcss', 'keep flush css files' ],
    [ '-l, --lazyload', 'disable hot reload' ],
    [ '-h, --https', 'start https server' ],
    [ '    --externals', 'webpack external varibles' ],
    [ '-o, --open', 'open url in default browser' ],
    [ '    --loose', 'use babel es2015 loose mode to transform codes' ],
    [ '    --historyApiFallback', 'history api fallback mappings' ],
    [ '    --mockapi', 'mock data api mappings' ],
    [ '    --includes', 'loader should include paths' ],
    [ '    --polyfill', 'use core-js to do polyfills' ],
    [ '    --alias', 'path alias' ],
    [ '    --injects', 'inject js into html' ],
    [ '    --no-sourcemap', 'don\'t use the `SourceMapDevToolPlugin` of plugins to debug' ],
  ],

  action: function(options) {

    // options
    var src = options.src;
    var dist = options.dist;
    var port = parseInt(options.port);
    var entry = options.entry;
    var pages = options.pages;
    var vars = options.vars;
    var buildvars = options.buildvars;
    var proxy = options.proxy;
    var keepcss = options.keepcss;
    var lazyload = options.lazyload;
    var httpsOpt = options.https;
    var loose = options.loose;
    var open = options.open;
    var externals = options.externals || {
      'react': 'window.React',
      'react-dom': 'window.ReactDOM || window.React'
    };
    var historyApiFallback = options.historyApiFallback;
    var mockapi = options.mockapi;
    var includes = options.includes;
    var polyfill = !!options.polyfill;
    var host =  options.host;
    var sourcemap = options.sourcemap;
    var alias = (function(aliasMap) {
      for (var key in aliasMap) {
        aliasMap[key] = util.cwdPath(src, aliasMap[key]);
      }
      return aliasMap;
    })(typeof options.alias === 'object' ? options.alias : {
      i18n: 'i18n'
    });
    var injects = options.injects || [];

    // find a usable ip address
    var ipAddr = host?host:ip.address();
    portscanner.findAPortNotInUse(port, port + 10, ipAddr, function(err, aPort) {
      if (err || !aPort) {
        console.error(chalk.red.bold('Port ' + port + ' in use. exit now!'));
        return process.exit(1);
      }
      if (aPort != port) {
        console.log(chalk.red.bold('Port ' + port + ' in use, Change to ' + aPort));
        port = aPort;
        open = true;
      }
      var address = (httpsOpt ? 'https' : 'http') + '://' + ipAddr + ':' + port;
      if (process.env.NOWA_UID) {
        util.writeProcessInfo({
          uid: process.env.NOWA_UID,
          address: address
        });
      }

      // entries
      var entries = {
        app: util.makeEntry({
          lazyload: lazyload,
          address: address
        }, src, entry)
      };

      // plugins
      var plugins = [
        new webpack.NoErrorsPlugin(),
      ];
      if (sourcemap) {
        plugins.push(
          new webpack.SourceMapDevToolPlugin({
            module: false
          })
        );
      }
      vars && plugins.push(new webpack.DefinePlugin(util.parseVars(vars)));
      !lazyload && plugins.push(new webpack.HotModuleReplacementPlugin());

      // compiler
      var resolveRoot = [
        util.relPath('..', 'node_modules')
      ];
      if (process.cwd() !== util.relPath('..', '..', '..')) {
        resolveRoot.push(util.relPath('..', '..'));
      }
      var compiler = preProcess({
        entry: pages ? util.makePageEntries({
          lazyload: lazyload,
          address: address
        }, src, entries, pages) : entries,
        output: {
          path: util.cwdPath(dist),
          filename: '[name]' + util.suffixByVars(vars, buildvars) + '.js',
          publicPath: '/'
        },
        plugins: plugins,
        resolve: {
          root: resolveRoot,
          alias: alias,
          extensions: ['', '.js', '.jsx']
        },
        resolveLoader: {
          root: resolveRoot
        },
        externals: externals,
        cache: true,
        module: {
          loaders: loader(options)
        }
      });
      var webpackCompiler = webpack(compiler);

      // dev server
      var app = express();

      // some pre-process
      app.use(function(req, res, next) {

        // deal with proxied request
        if (req.url.indexOf('/') !== 0) {
          req.url = '/' + req.url.split('/').slice(3).join('/');
        }

        // exclude css request with entry name
        if (!keepcss) {
          var match = /\/([^\/]+)\.css$/.exec(req.url);
          if (match && (match[1] in entries)) {
            res.setHeader('Content-Type', 'text/css; charset=utf-8');
            return res.end();
          }
        }

        // support cors
        res.setHeader('Access-Control-Allow-Origin', '*');

        next();
      });

      // favicon
      app.use(favicon(util.cwdPath('favicon.ico')));

      // dev middleware
      var devMiddlewareOpts = {
        publicPath: '/',
        stats: {
          chunks: false
        }
      };
      var devMiddleware = webpackDevMiddleware(webpackCompiler, devMiddlewareOpts);
      if (pages) {
        app.use(function() { // for hot replace middleware when multi-entries changed
          devMiddleware.apply(this, arguments);
        });
      } else {
        app.use(devMiddleware);
      }

      // hot middleware
      if (!lazyload) {
        var hotMiddleware = webpackHotMiddleware(webpackCompiler);
        if (pages) {
          app.use(function() { // for hot replace middleware when multi-entries changed
            hotMiddleware.apply(this, arguments);
          });
        } else {
          app.use(hotMiddleware);
        }
      }

      // mock data api
      if (mockapi) {
        for (let rule in mockapi) {
          app.use(rule, function(req, res, next) {
            var modulePath = require.resolve(util.cwdPath(mockapi[rule]));
            delete require.cache[modulePath];
            require(modulePath)(req, res, next);
          });
        }
      }

      // history api fallback rewrites
      if (historyApiFallback) {
        for (let rule in historyApiFallback) {
          app.use(rule, function(req, res, next) {
            var url = historyApiFallback[rule];
            if (req.params && url.indexOf('{{') !== -1) {

              // params replace
              url = url.replace(/\{\{([^}]+)\}\}/g, function(p, p1) {
                return req.params[p1] || '';
              });
            }
            req._parsedUrl.pathname = url;
            next();
          });
        }
      }

      // html injects server
      app.use(injectsHtml('html', injects));
      app.use(injectsHtml('.', injects));

      // static server
      app.use(serveStatic('html'));
      app.use(serveStatic(util.cwdPath(src, 'lib')));
      app.use(serveStatic(dist));
      app.use(serveStatic('.'));

      // http proxy
      var proxyMiddleware = function(proxy) {
        if (!/https?:\/\//.test(proxy)) {
          proxy = 'http://' + proxy;
        }
        var parsedUrl = url.parse(proxy);
        var pathPrefix = parsedUrl.path === '/' ? '' : parsedUrl.path;
        return httpProxy(parsedUrl.protocol + '//' + parsedUrl.host, {
          forwardPath: function(req) {
            return pathPrefix + url.parse(req.originalUrl).path;
          }
        });
      }
      if (typeof proxy === 'string') {
        app.use('*', proxyMiddleware(proxy));
      } else if (typeof proxy === 'object') {
        for (var pattern in proxy) {
          app.use(pattern, proxyMiddleware(proxy[pattern]));
        }
      }

      // create https server
      if (httpsOpt) {

        app = https.createServer(util.makeKeyCert(httpsOpt), app);
      }

      // listen for dev server
      app.listen(port, function(err) {
        if (err) {
          console.log(err);
          return;
        }
        console.log('Listening at ' + chalk.green.bold(address));

        // open url in default browser
        if (open) {
          util.open(address);
        }
      });

      if (pages) {

        // watching dir changes to update entry
        util.watch(util.cwdPath(src, 'pages'), function() {
          var newEntries = {
            app: util.makeEntry({
              lazyload: lazyload,
              address: address
            }, src, entry)
          };
          util.makePageEntries({
            lazyload: lazyload,
            address: address
          }, src, newEntries, pages);
          if (JSON.stringify(newEntries) !== JSON.stringify(compiler.entry)) {
            compiler.entry = newEntries;
            webpackCompiler = webpack(compiler);
            devMiddleware = webpackDevMiddleware(webpackCompiler, devMiddlewareOpts);
            if (!lazyload) {
              hotMiddleware = webpackHotMiddleware(webpackCompiler);
            }
          }
        });
      }
    });
  }
};

// compiler pre-process
function preProcess(config) {
  var newConfig;
  try {
    newConfig = require(util.cwdPath('webpack.config.js'))(config, webpack);
  } catch (e) {
  }
  return newConfig || config;
}
